### PR TITLE
Added Schedule Feature (Extended)

### LINKED ISSUE #445: (5 pts) (3 pt remaining) Office hour submission dialogue box, to add and manage events in schedule
https://docs.google.com/document/d/1p96J3TO9WePZLhm-Xia1GBCxkiYanyz-949q9PUm_Kc/edit?usp=sharing

### PR DESCRIPTION
**Pull request by limyer**
Closes #445 #527 #516 #451 #491 #446 

Extended From [UserViewsEnhancement](https://github.com/gordon-cs/gordon-360-ui/pull/577l)

[Corresponding api version is currently in develop](https://github.com/gordon-cs/gordon-360-api/commit/f4d8f1d3aa623a3cb44891915c9ec2957d4f4705)

**Parts done on this branch will be indicated with (UVE2)**

This is such a big update, so I will go one by one

## New Dependencies
Since there are some new dependencies, `npm install `is required

### react-big-calendar
- Main Calendar API to implement most of the features concerning the schedule.
[Click here for the documentation](http://intljusticemission.github.io/react-big-calendar/examples/index.html)

### moment.js
- Provides localization for the calendar and time computation. Luxon does similar job but it cannot be used as a localizer in the calendar.
[Click here for the documentation](https://momentjs.com/)

### react-timeago
- From the timestamp provided by the back-end, it enables us to check how much time has passed from the last update.
[Click here for the documentation](https://www.npmjs.com/package/react-timeago)

### @material-ui/pickers (UVE2)
- Used for visualization of start time and end time picking.
[Click here for the documentation](https://material-ui-pickers.dev/)

### @date-io/moment (UVE2)
- Localizer for the picker
[Click here for the documentation](https://libraries.io/npm/@date-io%2Fmoment)

## File Structure
This is how newly added and modified files are structured

### src/components/SchedulePanel
- The main component used on profile views (in ExpansionPanel form)

- Contains child components and manages their calls for actual function implementation.

- Uses myschedule and schedulecontrol service to fetch and update the schedule information.

#### /ScheduleCalendar

- Where Calendar API component is used.

- Uses schedule and myschedule service to fetch schedule events and merge them together.

- Also Calendar component provides information fired from selecting events and slots.

- Renamed from /CourseSchedule (UVE2)

#### /EditDescriptionDialog
- Dialog box that gets input to edit description.

- Parent component (if exists) will provide the existing description.

- The information is sent to the parent component after the submit.

#### /MyScheduleDialog
- Dialog box that gets input to add and edit customized events(called MySchedule)

- Add/Edit is distinguished by the prop of parent component "IsDoubleClick". If it is double click, it will receive the selected event and display the existing data to edit the form.

- Also validates the input so that it can only be submitted if it meets the requirements

1. Start time should be after 6 AM - nullifies when all day is checked
2. End time should be before 10PM - nullifies when all day is checked
3. Start time should not be later than End time - nullifies when all day is checked
4. At least one day of the week has to be checked
5. Description cannot be an empty string and cannot exceed 50 characters
6. Location cannot be an empty string and cannot exceed 50 characters

- The information is sent to the parent component after the submit.

- ~~It was named before the decision that confirmed customized events are no longer bound to office hours.~~
- Renamed from /OfficeHoursDialog (UVE2)

#### /RemoveScheduleDialog
- Dialog box that confirms if the user wants to delete the event.

- Only enabled when the event is not a course schedule.

- ~~It was named before the decision that confirmed customized events are no longer bound to office hours.~~
- Renamed from /RemoveHoursDialog (UVE2)
- The information is sent to the parent component after the submit.

### src/views
Modification made on Views

#### /Profile
- imports SchedulePanel with its profile property

- Locates the SchedulePanel with grid layout.

#### /MyProfile
- imports SchedulePanel with its profile property and indicate it is my profile

- Locates the SchedulePanel with grid layout

### src/services
Theses services corresponds api endpoints and communicates with the back-end.

#### /myschedule
- Gets myschedule data in following format.

```
/**
 * @global
 * @typedef MySchedule
 * @property {number} EVENT_ID
 * @property {String} LOCATION
 * @property {String} DESCRIPTION
 * @property {String} MON_CDE
 * @property {String} TUE_CDE
 * @property {String} WED_CDE
 * @property {String} THU_CDE
 * @property {String} FRI_CDE
 * @property {String} SAT_CDE
 * @property {String} SUN_CDE
 * @property {boolean} IS_ALLDAY
 * @property {TimeSpan} BEGIN_TIME
 * @property {TimeSpan} END_TIME
 */
```

- Furnishes this data to fit into the event format for the Calendar API.

- Add, update and delete myschedule.

#### /schedule
- Gets course schedule data in following format.

```

/**
 * @global
 * @typedef Schedule
 * @property {String} CRS_CDE
 * @property {String} CRS_TITLE
 * @property {String} BLDG_CDE
 * @property {String} ROOM_CDE
 * @property {String} MONDAY_CDE
 * @property {String} TUESDAY_CDE
 * @property {String} WEDNESDAY_CDE
 * @property {String} THURSDAY_CDE
 * @property {String} FRIDAY_CDE
 * @property {TimeSpan} BEGIN_TIME
 * @property {TimeSpan} END_TIME
 */

```

- Furnishes this data to fit into the event format for the Calendar API.

- It has two different "get" methods to differentiate myProfile and Profile => Privacy issue

#### /schedulecontrol
- Gets general schedule information in following format

```

/**
 * @global
 * @typedef ScheduleControl
 * @property {Number} IsSchedulePrivate
 * @property {DateTime} ModifiedTimeStamp
 * @property {String} Description
 */

```

- Updates privacy and description

- While updating description, it converts the string with some special characters to designated words because url parameter will not handle them properly. This has to be converted back when it comes back to the front-end. (SchedulePanel component does this)

- TimeStamp is updated on back-end when the modification request is sent.

### Calendar API
- Displays in weekly view (Actually, this is day view having multiple columns. Columns have their own id code as "ResourceId". The date is always set as today)

- Green line showing current time.

- Shows current month and year on top.

- Differentiates course schedule and customized schedule by color.

- Multiple events can be displayed on the same time.

- Hidden description can be shown when you hover your cursor on the event.

### Course Schedule
- Shows your overall semester course schedule in weekly view. (Currently displaying Fall 2018 schedule because of the test purpose. Should be changed on backend for production)

- Courses have their Course code and Room number.

### My Schedule
- Shows your customized weekly schedule

- You can add your own schedule by selecting slots on the calendar view. The time and the day of week is auto-filled according to your selection.

- You can edit the existing myschedule by double-clicking on the event. The existing information will be auto-filled.

- Selecting an event and pressing the remove button will remove the event from the schedule.

- The event can also be "All Day".

- You can use visualized time picker while picking start and end time (UVE2)

### Additional Information
- You can add and edit your own description or additional information on the schedule.

- Last update with your description and my schedule will be displayed as "Last Updated ... Ago"
_with different color (UVE2)_

### Privacy
- Students can hide their course schedules from other students. Other positions (except guest view) will be able to look into their schedules.

- Faculty schedule will always be displayed so faculty does not have privacy option.

- My Schedule will always be displayed no matter what.

## Known issues
- ~~In some cases, time is always invalid when someone tries to make an event because it does not handle the change between "AM" and "PM". It can be tempararily resolved by checking and unchecking all day option.~~ Fixed (UVE2)

- ~~Overall schedule is not responsive at all. Probably because of CSS grid layout that does not match with the existing design format.~~ Fixed (UVE2)

- ~~Having long description on AllDay schedule affects the location of the Calendar.~~ Fixed(UVE2)

- ~~Fixed : Now whole page is refreshed whenever we want to check any changes made.~~

- _.scss for CourseSchedule(Calendar API) is having extraneous styles those are not used for our calendar view. It is quite a challenge to figure out which one is used and which one is not._ Partially done (UVE2)

## Future plans
- Fix above issues
- Add Print as PDF feature
- View by Session
- ???

### CODE REVIEW

File: src/views/MyProfile/index.js [hunk 1]
```diff
@@ -724,6 +723,14 @@ class Profile extends Component {
                     </Card>
                   </Grid>
 
+                  <Grid item xs={12} lg={12} align="center">
+                <Grid container xs={12} lg={10} spacing="16" justify="center">
```
Comment (EjPlatzer) [inline src/views/MyProfile/index.js]: Can we indent these more?

File: src/views/Profile/index.js [hunk 2]
```diff
@@ -371,6 +373,14 @@ export default class Profile extends Component {
                     </CardContent>
                   </Card>
                 </Grid>
+                
+                <Grid item xs={12} lg={12} align="center">
+              <Grid container xs={12} lg={10} spacing="16" justify="center">
```
Comment (EjPlatzer) [inline src/views/Profile/index.js]: Also indent these more if we can?

File: src/views/MyProfile/index.js (hunk 1 above)
Comment (limyer) [reply src/views/MyProfile/index.js]: Change made

File: src/views/Profile/index.js (hunk 2 above)
Comment (limyer) [reply src/views/Profile/index.js]: Change made

**Pull request merged**